### DISCUSSION TITLE
Need help on table schema design for multi tenants

### DISCUSSION THREAD

**Discussion by jiayeZhu** [ClickHouse/ClickHouse#55580] [Q&A]
Hi,

We need to build some tables to support a lot of accounts (20K~40k), and we are storing weekly data for up to 20 years, so our first idea was
```sql
CREATE TABLE user_data_local ON CLUSTER '{cluster}' (
    account_id UInt64
    yearweek UInt64,
    col_001 UInt64,
    col_002 UInt64,
    data_001 Nullable(Float64),
    ...
    data_200 Nullable(Float64)
)
ENGINE = ReplicatedMergeTree(...)
PARTITION BY (account_id, yearweek)
ORDER BY (account_id, yearweek,col_001 , col_002)
SETTINGS  storage_policy='my_hdd_policy', max_parts_in_total=42000000; 
```
but then we found 2 issues:
1. insert speed becomes much slower when we have a lot of partitions (>1million)
2. restart clickhouse-server becomes very slow because of loading metadata.

we noticed [this solution](https://github.com/ClickHouse/ClickHouse/pull/32928). it reduced the restart time to <30s but it made the insert speed worse when inserting into multiple partitions at the same time.

so now looks like we have to change partition by (account_id) only so less partitions are needed and the insert speed looks good. and since yearweek is in primary key the query speed also looks good.
```sql
CREATE TABLE user_data_v2_local ON CLUSTER '{cluster}' (
    account_id UInt64
    yearweek UInt64,
    col_001 UInt64,
    col_002 UInt64,
    data_001 Nullable(Float64),
    ...
    data_200 Nullable(Float64)
)
ENGINE = ReplicatedMergeTree(...)
PARTITION BY (account_id)
ORDER BY (account_id, yearweek,col_001 , col_002)
SETTINGS  storage_policy='my_hdd_policy', use_metadata_cache = true; 
```

but we need to do reload on specific (account_id,yearwek) several times every day, then partition by account_id only becomes challenging... We tried the ReplacingMergeTree but the query performance with `FINAL` is not acceptable... any suggestions for this issue?

Thanks!

**Comment by jiayeZhu** [comment]
I just noticed that use_metadata_cache  is removed since 23.8 because of potential data loss.
then another question is how to speed up restart process when there are a lot of parts

Best Regards.

**Comment by royxact** [comment]
Another useful trick is not to partition by account_id directly,  but partition according to a hash:
Let's assume you plan to have 100K accounts. 
Let's design for X amount of partitions. Let's go for 512.
So you partition by the expression (account_id % 512).
That way you'll evenly split the accounts across partitions, while still controlling how many parts you have.
Clickhouse will anyway create a new part for each 1M rows, so if an account has a lot of data, it will anyway be split into many parts that contain data from that account_id.
As long as you keep ORDER BY account_id you should be fine, as clickhouse will skip parts which aren't relevant.

**Comment by den-crane** [comment]
There is no easy answer. 
I would consider 

```
PARTITION BY toYYYYMM(yearweek)
ORDER BY (account_id, yearweek,col_001 , col_002)
```

and using `delete` or `alter table delete` to remove data which should be reloaded.

**Comment by den-crane** [comment]
>We tried the ReplacingMergeTree but the query performance with FINAL is not acceptable... any suggestions for this issue?

In the latest Clickhouse there is an optimization which improves FINAL performance for fully optimized partitions https://github.com/ClickHouse/ClickHouse/pull/53919#issuecomment-1722286503

**Comment by den-crane** [comment]
Also if you are going to really store data for 20 years, with weekly partitioning it will be 52*20 = 1040 partitions, it also too much. Monthly is more reasonable.
